File: src/main/java/sharev/domain/gathering/service/GatheringService.java
package sharev.domain.gathering.service;

import java.util.List;
import java.util.Map;
import java.util.UUID;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import sharev.domain.gathering.dto.request.RequestCreateGatheringDto;
import sharev.domain.gathering.dto.request.RequestUpdateGatheringDto;
import sharev.domain.gathering.dto.response.ResponseGatheringDetailDto;
import sharev.domain.gathering.entity.Gathering;
import sharev.domain.gathering.entity.IntroduceTemplate;
import sharev.domain.gathering.entity.IntroduceTemplateContent;
import sharev.domain.gathering.exception.GatheringNotFoundException;
import sharev.domain.gathering.repository.GatheringRepository;
import sharev.domain.gathering.repository.IntroduceTemplateRepository;
import sharev.domain.team.entity.Team;
import sharev.domain.team.exception.TeamNotFoundException;
import sharev.domain.team.repository.TeamRepository;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class GatheringService {

    private final GatheringRepository gatheringRepository;
    private final IntroduceTemplateRepository introduceTemplateRepository;
    private final TeamRepository teamRepository;

    @Transactional
    public void create(Long teamId, RequestCreateGatheringDto dto) {
        Team team = teamRepository.findById(teamId)
                .orElseThrow(TeamNotFoundException::new);

        Gathering gathering = gatheringRepository.save(new Gathering(
                team, dto.visible(), dto.title(), dto.content(),
                dto.startAt(), dto.endAt(), dto.place(),
                dto.imageUrl(), dto.gatheringUrl(), dto.contact(),
                dto.registerStartAt(), dto.registerEndAt()));

        IntroduceTemplateContent templateContent = dto.introduceTemplate() != null
                ? dto.introduceTemplate()
                : new IntroduceTemplateContent("", Map.of());

        introduceTemplateRepository.save(new IntroduceTemplate(gathering, 0, templateContent));
    }

    public List<ResponseGatheringDetailDto> getGatherings(Long teamId) {
        Team team = teamRepository.findById(teamId)
                .orElseThrow(TeamNotFoundException::new);

        return gatheringRepository.findAllByTeam(team).stream()
                .map(ResponseGatheringDetailDto::new)
                .toList();
    }

    public ResponseGatheringDetailDto getGathering(Long teamId, UUID gatheringId) {
        Gathering gathering = getGatheringWithTeamValidation(teamId, gatheringId);
        return new ResponseGatheringDetailDto(gathering);
    }

    @Transactional
    public ResponseGatheringDetailDto update(Long teamId, UUID gatheringId, RequestUpdateGatheringDto dto) {
        Gathering gathering = getGatheringWithTeamValidation(teamId, gatheringId);

        gathering.update(dto.visible(), dto.title(), dto.content(),
                dto.startAt(), dto.endAt(), dto.place(),
                dto.imageUrl(), dto.gatheringUrl(), dto.contact(),
                dto.registerStartAt(), dto.registerEndAt());

        return new ResponseGatheringDetailDto(gathering);
    }

    @Transactional
    public void delete(Long teamId, UUID gatheringId) {
        Gathering gathering = getGatheringWithTeamValidation(teamId, gatheringId);
        gathering.softDelete();
    }

    private Gathering getGatheringWithTeamValidation(Long teamId, UUID gatheringId) {
        Gathering gathering = gatheringRepository.findById(gatheringId)
                .orElseThrow(GatheringNotFoundException::new);

        if (!gathering.getTeam().getId().equals(teamId)) {
            throw new GatheringNotFoundException();
        }

        return gathering;
    }
}

File: src/main/kotlin/sharev/gathering/application/service/GatheringService.kt
package sharev.gathering.application.service

import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import sharev.gathering.application.port.inbound.command.CreateGatheringCommand
import sharev.gathering.application.port.inbound.command.UpdateGatheringCommand
import sharev.gathering.application.port.inbound.mapper.toCreateGatheringResult
import sharev.gathering.application.port.inbound.mapper.toDetailResult
import sharev.gathering.application.port.inbound.mapper.toResult
import sharev.gathering.application.port.inbound.result.*
import sharev.gathering.application.port.inbound.usecase.*
import sharev.gathering.application.port.outbound.*
import sharev.gathering.domain.exception.GatheringException
import sharev.gathering.domain.exception.GatheringExceptionCode
import sharev.gathering.domain.model.Gathering
import sharev.team.application.port.outbound.CheckTeamMemberPort
import sharev.team.domain.exception.TeamException
import sharev.team.domain.exception.TeamExceptionCode
import java.util.*

@Service
@Transactional(readOnly = true)
class GatheringService(
    private val checkGatheringParticipantPort: CheckGatheringParticipantPort,
    private val saveGatheringPort: SaveGatheringPort,
    private val loadGatheringPort: LoadGatheringPort,
    private val loadIntroduceTemplatePort: LoadIntroduceTemplatePort,
    private val checkTeamMemberPort: CheckTeamMemberPort,
    private val loadParticipatedGatheringsPort: LoadParticipatedGatheringsPort,
) : CheckGatheringParticipantUseCase,
    CreateGatheringUseCase,
    GetTeamGatheringUseCase,
    UpdateGatheringUseCase,
    DeleteGatheringUseCase,
    GetIntroduceTemplateUseCase,
    GetParticipatedGatheringsUseCase,
    GetGatheringsUseCase {

    override fun isParticipant(accountId: Long, gatheringId: UUID): ParticipantResult {
        return ParticipantResult(checkGatheringParticipantPort.isParticipant(gatheringId, accountId))
    }

    @Transactional
    override fun create(command: CreateGatheringCommand): CreateGatheringResult {
        validateTeamAdmin(command.accountId, command.teamId)

        return saveGatheringPort.save(
            Gathering(
                id = Gathering.NEW_ID,
                teamId = command.teamId,
                visible = command.visible,
                title = command.title,
                content = command.content,
                startAt = command.startAt,
                endAt = command.endAt,
                place = command.place,
                imageUrl = command.imageUrl,
                gatheringUrl = command.gatheringUrl,
                contact = command.contact,
                registerStartAt = command.registerStartAt,
                registerEndAt = command.registerEndAt,
            ),
        ).toCreateGatheringResult()
    }

    override fun getParticipatedGatherings(accountId: Long, pageable: Pageable): Page<GatheringDetailResult> {
        return loadParticipatedGatheringsPort.loadParticipatedGatherings(accountId, pageable)
            .map { it.toDetailResult() }
    }

    override fun getGatherings(pageable: Pageable): Page<GatheringDetailResult> {
        return loadGatheringPort.loadAll(pageable)
            .map { it.toDetailResult() }
    }

    override fun getTeamGatherings(accountId: Long, teamId: Long): List<GatheringDetailResult> {
        validateTeamMember(accountId, teamId)

        return loadGatheringPort.loadAllByTeam(teamId)
            .map { it.toDetailResult() }
    }

    override fun getTeamGathering(accountId: Long, teamId: Long, gatheringId: UUID): GatheringDetailResult {
        validateTeamMember(accountId, teamId)

        val gathering = loadGatheringPort.load(gatheringId)

        if (gathering.teamId != teamId) {
            throw GatheringException(GatheringExceptionCode.GATHERING_NOT_FOUND)
        }

        return gathering.toDetailResult()
    }

    @Transactional
    override fun update(command: UpdateGatheringCommand): GatheringDetailResult {
        validateTeamAdmin(command.accountId, command.teamId)

        return saveGatheringPort.update(
            Gathering(
                id = command.gatheringId,
                teamId = command.teamId,
                visible = command.visible,
                title = command.title,
                content = command.content,
                startAt = command.startAt,
                endAt = command.endAt,
                place = command.place,
                imageUrl = command.imageUrl,
                gatheringUrl = command.gatheringUrl,
                contact = command.contact,
                registerStartAt = command.registerStartAt,
                registerEndAt = command.registerEndAt,
            )
        ).toDetailResult()
    }

    @Transactional
    override fun delete(accountId: Long, teamId: Long, gatheringId: UUID): DeleteGatheringResult {
        validateTeamAdmin(accountId, teamId)
        validateGatheringBelongsToTeam(teamId, gatheringId)

        saveGatheringPort.softDelete(gatheringId)
        return DeleteGatheringResult(gatheringId)
    }

    override fun getLatestTemplate(gatheringId: UUID, accountId: Long): IntroduceTemplateResult {
        if (!checkGatheringParticipantPort.isParticipant(gatheringId, accountId)) {
            throw GatheringException(GatheringExceptionCode.GATHERING_PARTICIPANT_NOT_FOUND)
        }

        return loadIntroduceTemplatePort.loadLatest(gatheringId)
            .toResult()
    }

    private fun validateGatheringBelongsToTeam(teamId: Long, gatheringId: UUID) {
        val gathering = loadGatheringPort.load(gatheringId)

        if (gathering.teamId != teamId) {
            throw GatheringException(GatheringExceptionCode.GATHERING_NOT_FOUND)
        }
    }

    private fun validateTeamMember(accountId: Long, teamId: Long) {
        if (!checkTeamMemberPort.isMember(accountId, teamId)) {
            throw TeamException(TeamExceptionCode.NOT_TEAM_MEMBER)
        }
    }

    private fun validateTeamAdmin(accountId: Long, teamId: Long) {
        if (!checkTeamMemberPort.isAdminMember(accountId, teamId)) {
            throw TeamException(TeamExceptionCode.NOT_TEAM_ADMIN_MEMBER)
        }
    }
}
